File: src/main/java/ch/epfl/dias/calcite/adapter/pelago/metadata/PelagoRelMdRowCount.java
package ch.epfl.dias.calcite.adapter.pelago.metadata;

import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.core.Aggregate;
import org.apache.calcite.rel.core.EquiJoin;
import org.apache.calcite.rel.core.Join;
import org.apache.calcite.rel.core.Project;
import org.apache.calcite.rel.metadata.BuiltInMetadata;
import org.apache.calcite.rel.metadata.ChainedRelMetadataProvider;
import org.apache.calcite.rel.metadata.CyclicMetadataException;
import org.apache.calcite.rel.metadata.MetadataDef;
import org.apache.calcite.rel.metadata.MetadataHandler;
import org.apache.calcite.rel.metadata.ReflectiveRelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMdRowCount;
import org.apache.calcite.rel.metadata.RelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMetadataQuery;
import org.apache.calcite.util.BuiltInMethod;

import com.google.common.collect.ImmutableList;

import ch.epfl.dias.calcite.adapter.pelago.PelagoDeviceCross;
import ch.epfl.dias.calcite.adapter.pelago.PelagoPack;
import ch.epfl.dias.calcite.adapter.pelago.PelagoRouter;
import ch.epfl.dias.calcite.adapter.pelago.PelagoTableScan;
import ch.epfl.dias.calcite.adapter.pelago.PelagoUnion;
import ch.epfl.dias.calcite.adapter.pelago.PelagoUnnest;
import ch.epfl.dias.calcite.adapter.pelago.PelagoUnpack;
import ch.epfl.dias.calcite.adapter.pelago.RelPacking;

public class PelagoRelMdRowCount implements MetadataHandler<BuiltInMetadata.RowCount> {
  private static final PelagoRelMdRowCount INSTANCE = new PelagoRelMdRowCount();

  public static final RelMetadataProvider SOURCE =
      ChainedRelMetadataProvider.of(
          ImmutableList.of(
              ReflectiveRelMetadataProvider.reflectiveSource(
                  BuiltInMethod.ROW_COUNT.method, PelagoRelMdRowCount.INSTANCE),
              RelMdRowCount.SOURCE));

  private final RelMdRowCount def = new RelMdRowCount();

  protected PelagoRelMdRowCount(){}

  @Override public MetadataDef<BuiltInMetadata.RowCount> getDef() {
    return BuiltInMetadata.RowCount.DEF;
  }

  public Double getRowCount(PelagoUnnest rel, RelMetadataQuery mq) {
    return rel.estimateRowCount(mq);
  }

  public Double getRowCount(PelagoPack rel, RelMetadataQuery mq) {
    return Math.ceil(mq.getRowCount(rel.getInput()) / (1024*1024.0));
  }

  public Double getRowCount(Aggregate rel, RelMetadataQuery mq) {
    if (rel.getGroupCount() == 0) return 1.0;
    return def.getRowCount(rel, mq); // groupby's are generally very selective
  }

  public Double getRowCount(PelagoUnpack rel, RelMetadataQuery mq) {
    return mq.getRowCount(rel.getInput()) * (1024*1024.0);
  }

  public Double getRowCount(PelagoRouter rel, RelMetadataQuery mq) {
    return rel.estimateRowCount(mq);//mq.getRowCount(rel.getInput()) / 2;
  }

  public Double getRowCount(PelagoUnion rel, RelMetadataQuery mq) {
    return mq.getRowCount(rel.getInput(0)) + mq.getRowCount(rel.getInput(1));
  }

  public Double getRowCount(PelagoDeviceCross rel, RelMetadataQuery mq) {
    return mq.getRowCount(rel.getInput());
  }

  public Double getRowCount(PelagoTableScan rel, RelMetadataQuery mq) {
    double rc = rel.getTable().getRowCount();
    if (rel.getTraitSet().containsIfApplicable(RelPacking.Packed)) rc /= 1024;
    return rc;
  }



  public Double getRowCount(Join rel, RelMetadataQuery mq) {
    Double leftRows = mq.getRowCount(rel.getLeft());
    Double rightRows = mq.getRowCount(rel.getRight());
    Double rows = Math.max(leftRows, rightRows);
    RelNode small = (leftRows < rightRows) ? rel.getLeft() : rel.getRight();
    Double sel = mq.getPercentageOriginalRows(small);
    if (sel == null) return rows;
    return rows * sel;
  }
}

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/metadata/PelagoRelMdRowCount.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package ch.epfl.dias.calcite.adapter.pelago.metadata

import ch.epfl.dias.calcite.adapter.pelago.costs.CostModel
import ch.epfl.dias.calcite.adapter.pelago.rel._
import ch.epfl.dias.calcite.adapter.pelago.traits.RelPacking
import com.google.common.collect.ImmutableList
import org.apache.calcite.rel.core.{Aggregate, Join}
import org.apache.calcite.rel.metadata._
import org.apache.calcite.util.BuiltInMethod

import java.util.Objects

object PelagoRelMdRowCount {
  private val INSTANCE = new PelagoRelMdRowCount
  val SOURCE: RelMetadataProvider = ChainedRelMetadataProvider.of(
    ImmutableList.of(
      ReflectiveRelMetadataProvider.reflectiveSource(
        BuiltInMethod.ROW_COUNT.method,
        PelagoRelMdRowCount.INSTANCE
      ),
      RelMdRowCount.SOURCE
    )
  )
}
class PelagoRelMdRowCount protected ()
    extends MetadataHandler[BuiltInMetadata.RowCount] {
  final private val `def` = new RelMdRowCount

  override def getDef: MetadataDef[BuiltInMetadata.RowCount] =
    BuiltInMetadata.RowCount.DEF

  def getRowCount(rel: PelagoUnnest, mq: RelMetadataQuery): java.lang.Double =
    rel.estimateRowCount(mq)

  def getRowCount(rel: PelagoPack, mq: RelMetadataQuery): java.lang.Double = { //    return Math.ceil(mq.getRowCount(rel.getInput()) / (1024*1024.0));
    mq.getRowCount(rel.getInput) / CostModel.blockSize
  }

  def getRowCount(rel: Aggregate, mq: RelMetadataQuery): java.lang.Double = {
    if (rel.getGroupCount == 0) return 1.0
    val groupKey = rel.getGroupSet
// rowCount is the cardinality of the group by columns
    val distinctRowCount = mq.getDistinctRowCount(rel.getInput, groupKey, null)
// groupby's are generally very selective
    Objects.requireNonNullElseGet(
      distinctRowCount,
      () => `def`.getRowCount(rel, mq) / 10
    )
  }

  def getRowCount(rel: PelagoUnpack, mq: RelMetadataQuery): java.lang.Double =
    mq.getRowCount(rel.getInput) * CostModel.blockSize

  def getRowCount(rel: PelagoRouter, mq: RelMetadataQuery): java.lang.Double =
    rel.estimateRowCount(mq) //mq.getRowCount(rel.getInput()) / 2;

  def getRowCount(rel: PelagoUnion, mq: RelMetadataQuery): java.lang.Double =
    mq.getRowCount(rel.getInput(0)) + mq.getRowCount(rel.getInput(1))

  def getRowCount(
      rel: PelagoDeviceCross,
      mq: RelMetadataQuery
  ): java.lang.Double =
    mq.getRowCount(rel.getInput)

  def getRowCount(rel: PelagoSplit, mq: RelMetadataQuery): java.lang.Double =
    mq.getRowCount(rel.getInput(0)) / 2

  def getRowCount(
      rel: PelagoTableScan,
      mq: RelMetadataQuery
  ): java.lang.Double = {
    var rc = rel.getTable.getRowCount
// FIXME: using CostModel.blockSize() breaks jo/in ordering in ssbm100
    if (rel.getTraitSet.containsIfApplicable(RelPacking.Packed))
      rc /= CostModel.blockSize
    rc
  }

  def getRowCount(rel: Join, mq: RelMetadataQuery): java.lang.Double = {
    val leftRows = mq.getRowCount(rel.getLeft)
    val rightRows = mq.getRowCount(rel.getRight)
    val rows = Math.max(leftRows, rightRows)
    val small =
      if (leftRows < rightRows) rel.getLeft
      else rel.getRight
    val sel = mq.getPercentageOriginalRows(small)
    if (sel == null) return rows
    rows * sel
  }
}
